### PR TITLE
Allow docker-push step to seemlessly push to a default registry

### PR DESCRIPTION
**Pull request by tjames221188**
We now try to infer the registry from the repository, previously the user would have to specify:

```yaml
registry: https://quay.io
repository: quay.io/user/app
```

If a user don't specify a registry and the repository contains another registry than the Docker Hub one we automatically use the v2 of the registry.

### CODE REVIEW

File: docker/docker.go [hunk 1]
```diff
@@ -838,15 +843,34 @@ func (s *DockerPushStep) InitEnv(env *util.Environment) {
...
 		opts.Password = s.options.AuthToken
-		s.repository = fmt.Sprintf("%s/%s/%s", s.options.WerckerContainerRegistry.Host, s.options.ApplicationOwnerName, s.options.ApplicationName)
 		s.builtInPush = true
 	}
-	auther, _ := dockerauth.GetRegistryAuthenticator(opts)
 
+	return opts
+}
+
+// The IStep Interfacew
```
Comment (faiq) [inline docker/docker.go]: typo

Comment (hatchan) [inline docker/docker.go]: Looks like this can be removed

File: docker/docker.go [hunk 2]
```diff
@@ -43,6 +43,8 @@ import (
 	"github.com/wercker/wercker/core"
 	"github.com/wercker/wercker/util"
 	"golang.org/x/net/context"
+	"github.com/docker/distribution/reference"
+	"net/url"
```
Comment (hatchan) [inline docker/docker.go]: Can you move the stdlib imports to the list above. I recommend using goimports which will do these things for you.

File: docker/docker_push_test.go [hunk 3]
```diff
@@ -7,6 +7,7 @@ import (
 	"github.com/stretchr/testify/suite"
 	"github.com/wercker/wercker/core"
 	"github.com/wercker/wercker/util"
+	"fmt"
```
Comment (hatchan) [inline docker/docker_push_test.go]: Move to the list above

File: docker/docker.go (hunk 1 above)
Comment (tjames221188) [inline docker/docker.go]: Typo fixed. @hatchan I actually found the comment useful - it's nice to know that everything below this line implements the `Step` interface. Unless you strongly disagree, I'd like to leave it in.

File: docker/docker.go (hunk 2 above)
Comment (tjames221188) [inline docker/docker.go]: done

File: docker/docker_push_test.go (hunk 3 above)
Comment (tjames221188) [inline docker/docker_push_test.go]: done

File: docker/docker.go (hunk 1 above)
Comment (hatchan) [inline docker/docker.go]: Thought it was a comment for something which was removed. If it applies to everything below, then I don't mind.

File: docker/docker.go [hunk 4]
```diff
@@ -838,15 +842,34 @@ func (s *DockerPushStep) InitEnv(env *util.Environment) {
...
+		domain := reference.Domain(x)
+
+		if domain != "docker.io" {
+			reg := &url.URL{Scheme: "https", Host: domain, Path: "/v2"}
+			opts.Registry = reg.String() + "/"
+		}
+	}
+
+	// Set user and password automatically if using wercker registry
+	if opts.Registry == s.options.WerckerContainerRegistry.String()+"/v2/" {
```
Comment (jwolski2) [inline docker/docker.go]: @kokaz and I noticed a bug here. "/v2" must not be appended to the WerckerContainerRegistry value. @kokaz fixed this bug in another commit and will cherry-pick that commit into this branch.

**Comment by kokaz** [comment]
PR updated see 571ef3d

**Pull request merged**